clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -858,6 +885,12 @@ public void onSuccess(ClientResponse resp, RequestFuture<Void> future) {
         public void onFailure(RuntimeException e, RequestFuture<Void> future) {
             log.debug("FindCoordinator request failed due to {}", e.toString());
 
+            if (e instanceof UnsupportedBatchLookupException) {
```

Comment:
Is there some reason why we can't pass the `ApiVersions` into the coordinator, so we can get the batching right without needing to retry like this?

Reply:
I've only taken a very brief look and I think this approach would work well for Connect, Producer and Consumer, however it's a bit more complicated with Admin.

In Admin, requests are built by lookup strategies. Lookups can be sent to any broker so knowing the max version for a specific call is not completely trivial. That said, it's not impossible either so if there's concensus it would be preferable I can give that a try.

Reply:
This `AdminApiDriver` abstraction is pretty new to me, so I might be wide of the mark, but it doesn't appear to handle this very well. The lookup strategy has to build a request without knowing either the broker or the API versions. It would be possible to pass the `ApiVersions` to the `CoordinatorStrategy`, which should let you do the right thing based on the minimum of the FindCoordinator API version supported in the whole cluster. (That's not completely perfect, since really you'd want to decide on a per-broker basis, but I think it would be good enough). Sadly it's not quite enough to pass just `ApiVersions`, since it doesn't really know about the nodes in the cluster, so you'd need to pass `Metadata` too, which is quite a lot of work. So I can understand why it makes sense to do it like this, since it lets you benefit from the existing logic for figuring out request versions.

Reply:
@tombentley That's a really good point, perhaps we can file a JIRA in the future to add support to the `AdminApiDrive` to support passing in versions for different API calls. I'm currently working on KIP-709 where I'm trying to introduce batching to the fetch offsets API, and I'll need to do something similar as Mickael has done here to be able to support all versions, but having some native support to be able to know the version beforehand would be much more desirable.